linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,343 @@
+                result->srw_lock = srw_lock_create(false, "");
+                if (result->srw_lock == NULL)
+                {
+                    LogError("Failure srw_lock_create");
+                }
+                else
+                {
+                    if (sem_init(&result->semaphore , 0 , 0) != 0)
+                    {
+                        LogError("Failure creating semi_init");
+                    }
+                    else
+                    {
+                        result->thread_count = 0;
+                        result->task_count = 0;
+                        (void)interlocked_exchange(&result->quitting, 0);
+
+                        result->task_list = NULL;
+                        result->task_list_last = NULL;
+                        goto all_ok;
+                    }
+                    srw_lock_destroy(result->srw_lock);
+                }
+                free(result->thread_handle_list);
+            }
+            free(result);
+            result = NULL;
+        }
+    }
+all_ok:
+    return result;
+}
+
+void threadpool_destroy(THREADPOOL_HANDLE threadpool)
+{
+    if (threadpool == NULL)
+    {
+        // do nothing
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p", threadpool);
+    }
+    else
+    {
+        interlocked_exchange(&threadpool->quitting, 1);
+
+        for (size_t index = 0; index < threadpool->list_index; index++)
+        {
+            int dont_care;
+            if (ThreadAPI_Join(threadpool->thread_handle_list[index], &dont_care) != THREADAPI_OK)
+            {
+                LogError("Failure joining thread");
+            }
+        }
+
+        // Loop through
+        while (threadpool->task_list != NULL)
+        {
+            THREADPOOL_TASK* temp_item = threadpool->task_list;
+            threadpool->task_list = temp_item->next;
+            free(temp_item);
+        }
+
+        free(threadpool->thread_handle_list);
+
+        sem_destroy(&threadpool->semaphore);
+        srw_lock_destroy(threadpool->srw_lock);
+        free(threadpool);
+    }
+}
+
+int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCTION work_function, void* work_function_context)
+{
+    int result;
+    if (threadpool == NULL || work_function == NULL)
+    {
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p, THREADPOOL_TASK_FUNC work_function: %p", threadpool, work_function);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        THREADPOOL_TASK* task_item = malloc(sizeof(THREADPOOL_TASK));
+        if (task_item == NULL)
+        {
+            LogError("Failure allocating THREADPOOL_TASK structure");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            task_item->next = NULL;
+            task_item->task_param = work_function_context;
+            task_item->task_func = work_function;
+
+            {
+                srw_lock_acquire_exclusive(threadpool->srw_lock);
```

Comment:
>srw_lock_acquire_exclusive [](http://example.com/codeflow?start=16&length=26)

I think this can be done without this lock.

Reply:
Don't want to sidetrack. I looked around and there seem to be 2 threadpools out of the box:
- [https://github.com/Pithikos/C-Thread-Pool](https://github.com/Pithikos/C-Thread-Pool)
- [https://linux.die.net/man/3/cp_thread_pool](https://linux.die.net/man/3/cp_thread_pool)

Reply:
What I means when I said we can do it without the lock is:
We could have a singly linked list. Add items in it with interlocked.
Remove items and execute from it - since there is no guarantee of execution order for a threadpool.

Reply:
We can talk about it tomorrow.